src/main/java/studentlist/StudentList.java

```diff
@@ -53,47 +114,160 @@ public int getNumStudents() {
+
+        float gpa = validateGpa(parts[2].trim(), errorMessages);
+
+        ArrayList<Integer> preferences = validatePreferences(parts[3].trim(), errorMessages);
+
+        // If any errors occurred, throw a single exception with all error messages
+        if (!errorMessages.isEmpty()) {
+            throw new SEPException(String.join("\n", errorMessages));
         }
 
-        // Trim spaces and braces, then split the preferences
-        String preferencesData = parts[3].trim().replaceAll("[{}]", "");
-        String[] numberStrings = preferencesData.split(",");
+        return new Student(studentId, gpa, preferences);
+    }
 
-        ArrayList<Integer> preferences = new ArrayList<>();
-        for (String numberString : numberStrings) {
-            preferences.add(Integer.parseInt(numberString.trim()));
+    public void printStudentList() {
+        try {
+            ui.printStudentList(this.students);
+        } catch (SEPException e) {
+            ui.printResponse(e.getMessage());
         }
+    }
 
-        // Create and return the Student object
-        return new Student(parts[1].trim(), Float.parseFloat(parts[2].trim()), preferences);
+    public void generateReport() {
+        try {
+            ui.generateReport(students);
+        } catch (SEPException e) {
+            ui.printResponse(e.getMessage());
+        }
     }
 
+    /**
+     * Updates the current student list with a new one.
+     *
+     * @param studentList The new student list to set.
+     */
+    public void setStudentList(StudentList studentList) {
+        this.students = studentList.students; // Overwrite the current ArrayList
+    }
 
-    public void addStudent(Student student) {
-        students.add(student);
+    /**
+     * Organizes the student ID by removing all spaces,
+     * including any spaces in between the numbers or alphabets,
+     * before validating the student ID.
+     *
+     * @param id The raw student ID string.
+     * @return A trimmed and space-free student ID.
+     */
+    private String organiseId(String id) {
+        String studentId = id.trim();
+        return studentId.replaceAll("\\s+", "");
     }
 
-    public void deleteStudent(String id) {
-        String[] parts = id.split(" ");
-        students.removeIf(student -> student.getId().equals(parts[1]));
+    /**
+     * Splits the input string into parts based on the expected format
+     * and order: id/, gpa/, followed by p/. Any other order would be
+     * rejected and a format exception would be thrown.
+     *
+     * @param input The raw input string.
+     * @return A String array containing the parts of the input.
+     * @throws SEPFormatException If the input format or order is invalid.
+     */
+    private String[] splitInput(String input) throws SEPException {
+        // Updated regex to allow spaces in ID, GPA, and preferences
+        String pattern = "^add\\s+id/[\\S ]+\\s+gpa/[\\S ]+\\s+p/\\{[\\S ]+}$";
```

Comment:
Consider extracting the RegEx pattern into a static variable so that it's intuitive for others.

Reply:
fixed.